### DISCUSSION TITLE
Support requests-http-signature

### DISCUSSION THREAD

**Discussion by max-tet** [encode/httpx#2408] [Potential Issue] [answered]
HTTPX tries to be compatible with the requests package, including the use of callables for custom authentication. However, it is incompatible with [requests_http_signature](https://github.com/pyauth/requests-http-signature). When trying to sign a request, this is the error: `AttributeError: 'Request' object has no attribute 'body'`

requests_http_signature expects the `body` attribute [here](https://github.com/pyauth/requests-http-signature/blob/develop/requests_http_signature/__init__.py#L129), but the [Request](https://www.python-httpx.org/api/#request) object does not have the attribute because it is called `content`.

What I don't understand is that the [Request object of the Requests library](https://requests.readthedocs.io/en/latest/api/#requests.Request) also seems to have no attribute called `body`. Why does it work here then?

**Comment by lovelydinosaur** [comment, marked as answer]
> What I don't understand is that the [Request object of the Requests library](https://requests.readthedocs.io/en/latest/api/#requests.Request) also seems to have no attribute called body

It doesn't have a `body` parameter in the constructor. It does have [a body attribute on the instance](https://github.com/psf/requests/blob/7104ad4b135daab0ed19d8e41bd469874702342b/requests/models.py#L346).

(I can't find that documented, but I can see it in the source.)

> HTTPX tries to be compatible with the requests package, including the use of callables for custom authentication.

"[broadly compatible](https://www.python-httpx.org/compatibility/)". You can't use `requests` auth classes interchangeably with `httpx`, no.

**Reply by max-tet** [reply]
Thanks for the explanation. Seems like making the two packages compatible is [not as easy as I thought](https://github.com/pyauth/requests-http-signature/issues/37). I'll have to stick to Requests then.
